e2e/services/electron.ts

```diff
@@ -0,0 +1,100 @@
+MIT License
+
+Copyright (c) 2021-present, Elastic NV
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+import { _electron, ElectronApplication, Page } from "playwright";
+import { TestBrowserService } from "./browser";
+import path from "path";
+import { env } from ".";
+
+export class ElectronServiceFactory {
+  #instance: ElectronApplication;
+  #recordingBrowserPage: Page;
+
+  async getInstance() {
+    if (this.#instance) return this.#instance;
+
+    try {
+      this.#instance = await _electron.launch({
+        args: [path.join(__dirname, "../..", "electron", "electron.js")],
+        env: {
+          TEST_PORT: env.TEST_PORT,
+          PW_DEBUG: "console",
+          NODE_ENV: process.env.NODE_ENV,
+        },
+      });
+
+      return this.#instance;
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.log(e);
+    }
+  }
+
+  async getWindow() {
+    if (!this.#instance) await this.getInstance();
+    return this.#instance.firstWindow();
+  }
+
+  async terminate() {
+    if (!this.#instance) return;
+    await this.#instance.close();
+    this.#instance = null;
+  }
+
+  async enterTestUrl(testUrl: string) {
+    const electronWindow = await this.getWindow();
+    return await electronWindow.type(
+      '[placeholder="Enter URL to test"]',
+      testUrl
+    );
+  }
+
+  async clickStartRecording(waitUntilIdle = true) {
+    const electronWindow = await this.getWindow();
+    await electronWindow.click("text=Start recording");
+    this.#recordingBrowserPage =
+      await TestBrowserService.getRemoteBrowserPage();
+    if (waitUntilIdle) {
+      await this.waitForPageToBeIdle();
+    }
+    return this.#recordingBrowserPage;
+  }
+
+  async waitForPageToBeIdle(timeout = 45000) {
+    await this.#recordingBrowserPage.waitForLoadState("networkidle", {
+      timeout,
+    });
+  }
+
+  async navigateRecordingBrowser(url, waitUntilIdle = true, timeout = 4500) {
+    await this.#recordingBrowserPage.goto(url, { timeout });
```

Comment:
wait Until here and waitIdle on the next line is so different in terms of load.

1. this.#recordingBrowserPage.goto(url, { timeout });. You can pass waitUntil: "networkidle" in this and it would wait until the URL loaded has done loading all the network requests.

2. The next line `waitForPageToBeIdle` does not have any context on what to wait on as the page state would have already been loaded and wont be based on any actions. 

I would squeeze them in and have it as part of line 91 `wait this.#recordingBrowserPage.goto(url, { waitUntil: "load"/"networkidle", timeout });